lib/SILOptimizer/Transforms/RetainReleaseCodeMotion.cpp

```diff
@@ -0,0 +1,1073 @@
+  }
+}
+
+void RetainCodeMotionContext::initializeCodeMotionBBMaxSet() {
+  for (SILBasicBlock *BB : PO->getReversePostOrder()) {
+    // If basic block has no predecessor, do nothing.
+    BlockState *State = BlockStates[BB];
+    if (BB->pred_empty()) {
+      State->BBMaxSet.reset();
+    } else {
+      // Intersect in all predecessors' BBSetOut.
+      State->BBMaxSet.set();
+      for (auto E = BB->pred_end(), I = BB->pred_begin(); I != E; ++I) {
+        State->BBMaxSet &= BlockStates[*I]->BBMaxSet;
+     }
+   }
+
+   // Process the instructions in the basic block to find what refcounted
+   // roots are retained. If we know that a rc root cant be retained at a
+   // basic block, then we know we do not need to consider it for the killset.
+   // NOTE: this is a conservative approximation, because some retains may be
+   // blocked before it reaches this block.
+   for (auto &II : *BB) {
+      if (!isRetainInstruction(&II))
+        continue;
+      State->BBMaxSet.set(RCRootIndex[getRCRoot(&II)]);
+    }
+  }
+}
+
+void RetainCodeMotionContext::computeCodeMotionGenKillSet() {
+  for (SILBasicBlock *BB : PO->getReversePostOrder()) {
+    auto *State = BlockStates[BB];
+    bool InterestBlock = false; 
+    for (auto &I : *BB) {
+      // Check whether this instruction blocks any RC root code motion.
+      for (unsigned i = 0; i < RCRootVault.size(); ++i) {
+        if (!State->BBMaxSet.test(i) || !mayBlockCodeMotion(&I, RCRootVault[i]))
+          continue;
+        // This is a blocking instruction for the rcroot.
+        InterestBlock = true;
+        State->BBKillSet.set(i);
+        State->BBGenSet.reset(i);
+      }
+      // If this is a retain instruction, it also generates.
+      if (isRetainInstruction(&I)) {
+        unsigned idx = RCRootIndex[getRCRoot(&I)];
+        State->BBGenSet.set(idx);
+        assert(State->BBKillSet.test(idx) && "Killset computed incorrectly");
+        State->BBKillSet.reset(idx);
+        InterestBlock = true;
+      }
+    }
+
+    // Is this a block thats interesting to the last iteration of the data flow.
+    if (!InterestBlock)
+      continue;
+    InterestBlocks.insert(BB);
+  }
+}
+
+bool RetainCodeMotionContext::performCodeMotion() {
+  bool Changed = false;
+  // Create the new retain instructions.
+  for (auto RC : InsertPoints) {
+    for (auto IP : RC.second) {
+      createIncrementBefore(RC.first, IP);
+      Changed = true;
+    }
+  }
+  // Remove the old retain instructions.
+  for (auto R : RCInstructions) {
+    ++NumRetainsSunk;
+    recursivelyDeleteTriviallyDeadInstructions(R, true);
+  }
+  return Changed;
+}
+
+void RetainCodeMotionContext::mergeBBDataFlowStates(SILBasicBlock *BB) {
+  BlockState *State = BlockStates[BB];
+  State->BBSetIn.reset();
+  // If basic block has no predecessor, simply reset and return.
+  if (BB->pred_empty())
+    return;
+
+  // Intersect in all predecessors' BBSetOuts.
+  State->BBSetIn.set();
```

Comment:
You should initialize this with the first predecessor's out set to save from having to set all the bits and then immediately clear many of them.